gateway/service/src/main/java/org/eclipse/ditto/gateway/service/endpoints/actors/CheckPermissionsActor.java

```diff
@@ -0,0 +1,417 @@
+     * @param permissionResults the map to store the permission check results.
+     */
+    private void processPolicyCommandResult(final Object result,
+            final Map<String, ResourcePermissions> resourcePermissions,
+            final Map<String, Boolean> permissionResults) {
+        if (result instanceof CheckPolicyPermissionsResponse response) {
+            permissionResults.putAll(CheckPolicyPermissionsResponse.toMap(response.getPermissionsResults()));
+        } else {
+            setPermissionCheckFailure(resourcePermissions, permissionResults);
+        }
+    }
+
+    /**
+     * Marks the permission check as failed by setting all results for the given resources to {@code false}.
+     *
+     * @param resourcePermissions the map of resource permissions that failed.
+     * @param permissionResults the map of permission results where the failure should be reflected.
+     */
+    private void setPermissionCheckFailure(Map<String, ResourcePermissions> resourcePermissions,
+            Map<String, Boolean> permissionResults) {
+        resourcePermissions.forEach((key, value) -> permissionResults.put(key, false));
+    }
+
+    /**
+     * Handles a failure during the permission checking process by logging the error and sending a
+     * {@link DittoInternalErrorException} back to the sender.
+     *
+     * @param ex the exception that occurred during the permission check.
+     * @param command the {@link CheckPermissions} command containing permission checks.
+     * @return null, as the error is propagated via an exception and response is sent.
+     */
+    private Void handleFailure(Throwable ex, CheckPermissions command) {
+        logError(ex, command.getPermissionChecks(), command.getDittoHeaders());
+
+        final DittoInternalErrorException errorResponse = DittoInternalErrorException.newBuilder()
+                .dittoHeaders(command.getDittoHeaders())
+                .cause(ex)
+                .message("Internal error while processing permission checks for command: " + command)
+                .build();
+        sender.tell(errorResponse, getSelf());
+        throw new CompletionException(ex);
+    }
+
+
+    /**
+     * Logs an error message when a permission check fails for a given request.
+     * <p>
+     * This method logs the failure of a permission check operation using a {@code logger} with correlation ID
+     * extracted from the provided {@code DittoHeaders}. The logged message includes the details of the failed
+     * permission check request and the associated exception message.
+     * </p>
+     *
+     * @param ex the exception that occurred during the permission check.
+     * @param permissionChecks the map containing the permission checks that failed.
+     * @param headers the {@code DittoHeaders} associated with the command, used for logging context.
+     */
+    private void logError(Throwable ex, LinkedHashMap<String, ImmutablePermissionCheck> permissionChecks,
```

Comment:
params should be `final`
Instead of `LinkedHashMap` as param, you should use `Map` - `LinkedHashMap` is just the implementation we choose.